File: apps/web/lib/api/channels.ts
import { request } from "@/lib/http";

/** 后端统一响应信封（见 movieclaw_api.schemas.response.ApiResponse） */
interface ApiEnvelope<T> {
  success: boolean;
  code: string;
  message: string;
  data: T;
}

async function unwrap<T>(promise: Promise<ApiEnvelope<T>>): Promise<T> {
  return (await promise).data;
}

/** 已绑定的微信账号（见 schemas.channels.WeixinAccountView）。 */
export interface WeixinAccount {
  account_id: string;
  /** 扫码绑定人的微信用户 id（白名单：只有此人能对话） */
  bound_user_id: string | null;
  /** active=正常；stale=凭据失效需重新扫码 */
  status: "active" | "stale";
  /** 当前进程内收发循环是否在运行 */
  running: boolean;
  last_error: string | null;
  bound_at: string;
}

/** 绑定流程状态（见 schemas.channels.WeixinBindingStatusView 的注释）。 */
export type WeixinBindingStatus =
  | "pending"
  | "scanned"
  | "need_verify_code"
  | "confirmed"
  | "already_bound"
  | "expired"
  | "failed";

export interface WeixinBindingStart {
  challenge_id: string;
  qrcode_url: string;
  /** 服务端渲染好的二维码 SVG data URL，<img> 直接显示 */
  qrcode_image: string;
  message: string;
}

export interface WeixinBindingSnapshot {
  challenge_id: string;
  status: WeixinBindingStatus;
  message: string;
  qrcode_url: string;
  qrcode_image: string;
  account: WeixinAccount | null;
}

export function listWeixinAccounts(init?: RequestInit): Promise<WeixinAccount[]> {
  return unwrap(request<ApiEnvelope<WeixinAccount[]>>("/channels/weixin/accounts", init));
}

export function startWeixinBinding(): Promise<WeixinBindingStart> {
  return unwrap(
    request<ApiEnvelope<WeixinBindingStart>>("/channels/weixin/bindings", { method: "POST" }),
  );
}

export function getWeixinBindingStatus(challengeId: string): Promise<WeixinBindingSnapshot> {
  return unwrap(
    request<ApiEnvelope<WeixinBindingSnapshot>>(
      `/channels/weixin/bindings/${encodeURIComponent(challengeId)}`,
    ),
  );
}

export function submitWeixinVerifyCode(
  challengeId: string,
  code: string,
): Promise<Record<string, never>> {
  return unwrap(
    request<ApiEnvelope<Record<string, never>>>(
      `/channels/weixin/bindings/${encodeURIComponent(challengeId)}/verify-code`,
      { method: "POST", body: JSON.stringify({ code }) },
    ),
  );
}

export function unbindWeixinAccount(accountId: string): Promise<Record<string, never>> {
  return unwrap(
    request<ApiEnvelope<Record<string, never>>>(
      `/channels/weixin/accounts/${encodeURIComponent(accountId)}`,
      { method: "DELETE" },
    ),
  );
}

File: src/movieclaw_api/schemas/channels.py
"""IM 通道(微信绑定)相关的请求/响应模型。"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from movieclaw_db.models.channel_account import ChannelAccount


class WeixinAccountView(BaseModel):
    """已绑定账号(绑定页列表项)。"""

    account_id: str
    #: 扫码绑定人的微信用户 id(白名单)
    bound_user_id: str | None
    #: active=正常;stale=凭据失效,需重新扫码
    status: str
    #: 当前进程内收发循环是否在运行
    running: bool
    last_error: str | None
    bound_at: datetime

    @classmethod
    def from_model(cls, row: ChannelAccount, *, running: bool) -> WeixinAccountView:
        return cls(
            account_id=row.account_id,
            bound_user_id=row.bound_user_id,
            status=row.status,
            running=running,
            last_error=row.last_error,
            bound_at=row.created_at,
        )


class WeixinBindingStartView(BaseModel):
    """发起绑定的返回:前端渲染二维码并开始 poll。"""

    challenge_id: str
    #: 二维码内容链接(前端用它渲染二维码图)
    qrcode_url: str
    message: str


class WeixinBindingStatusView(BaseModel):
    """绑定状态快照(前端每 1-2 秒 poll 一次,后端只读内存,毫秒级返回)。

    status 取值:
    pending / scanned / need_verify_code / confirmed / already_bound /
    expired / failed。confirmed 时通道已启动,account 字段附带账号信息。
    """

    challenge_id: str
    status: str
    message: str
    #: 二维码可能中途刷新(过期自动换新),前端每次 poll 都以此为准重绘
    qrcode_url: str
    account: WeixinAccountView | None = None


class WeixinVerifyCodePayload(BaseModel):
    """提交手机微信上显示的配对数字。"""

    code: str = Field(min_length=1, max_length=16, description="配对码")
